File: src/main/java/net/ndrei/teslacorelib/gui/WorkEnergyIndicatorPiece.java
package net.ndrei.teslacorelib.gui;

import com.google.common.collect.Lists;
import com.mojang.realmsclient.gui.ChatFormatting;
import net.ndrei.teslacorelib.tileentities.IWorkEnergyProvider;

import java.util.List;

/**
 * Created by CF on 2016-12-27.
 */
public class WorkEnergyIndicatorPiece extends BasicContainerGuiPiece {
    private IWorkEnergyProvider provider;

    public WorkEnergyIndicatorPiece(IWorkEnergyProvider provider, int left, int top) {
        super(left, top, 36, 4);

        this.provider = provider;
    }

    @Override
    public void drawBackgroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, float partialTicks, int mouseX, int mouseY) {
        container.bindDefaultTexture();

        container.drawTexturedModalRect(guiX + this.getLeft(), guiY + this.getTop(), 1, 245, this.getWidth(), this.getHeight());
        if (this.provider != null) {
            float percent = (float)this.provider.getWorkEnergyStored() / (float)this.provider.getWorkEnergyCapacity();
            int width = Math.max(0, Math.min(this.getWidth() - 2, Math.round((this.getWidth() - 2) * percent)));
            if (width > 0) {
                container.drawTexturedModalRect(
                        guiX + this.getLeft() + 1, guiY + this.getTop() + 1,
                        2, 251, width, this.getHeight() - 2);
            }
        }
    }

    @Override
    public void drawForegroundTopLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
        if (super.isInside(container, mouseX, mouseY) && (this.provider != null)) {
            List<String> lines = Lists.newArrayList();
            lines.add(String.format("%sWork Energy Buffer", ChatFormatting.DARK_PURPLE));
            lines.add(String.format("%s%,d T %sof %s%,d T",
                    ChatFormatting.AQUA, this.provider.getWorkEnergyStored(),
                    ChatFormatting.DARK_GRAY,
                    ChatFormatting.RESET, this.provider.getWorkEnergyCapacity()));
            lines.add(String.format("%smax: %s+%,d T %s/ tick",
                    ChatFormatting.GRAY,
                    ChatFormatting.AQUA, this.provider.getWorkEnergyTick(),
                    ChatFormatting.GRAY));

            container.drawTooltip(lines, mouseX - guiX, mouseY - guiY);
        }
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/gui/WorkEnergyIndicatorPiece.kt
package net.ndrei.teslacorelib.gui

import net.minecraft.util.text.TextFormatting
import net.ndrei.teslacorelib.MOD_ID
import net.ndrei.teslacorelib.config.TeslaCoreLibConfig
import net.ndrei.teslacorelib.localization.GUI_WORK_ENERGY
import net.ndrei.teslacorelib.localization.localizeModString
import net.ndrei.teslacorelib.localization.makeTextComponent

/**
 * Created by CF on 2017-06-28.
 */
class WorkEnergyIndicatorPiece(private val provider: IWorkEnergyProvider?, left: Int, top: Int)
    : BasicContainerGuiPiece(left, top, 36, 4) {
    private val displayType get() = TeslaCoreLibConfig.energyDisplay

    override fun drawBackgroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {
        container.bindDefaultTexture()

        container.drawTexturedModalRect(guiX + this.left, guiY + this.top, 1, 245, this.width, this.height)
        if (this.provider != null) {
            val percent = this.provider.workEnergyStored.toFloat() / this.provider.workEnergyCapacity.toFloat()
            val width = Math.max(0, Math.min(this.width - 2, Math.round((this.width - 2) * percent)))
            if (width > 0) {
                container.drawTexturedModalRect(
                        guiX + this.left + 1, guiY + this.top + 1,
                        this.displayType.workIcon.left, this.displayType.workIcon.top, width, this.height - 2)
            }
        }
    }

    override fun drawForegroundTopLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {
        if (super.isInside(container, mouseX, mouseY) && this.provider != null) {
            val energySystem = this.displayType
            val lines = (if (this@WorkEnergyIndicatorPiece.provider.hasWorkBuffer) mutableListOf(
                localizeModString(MOD_ID, GUI_WORK_ENERGY, "work energy buffer") {
                    +TextFormatting.DARK_PURPLE
                },
                localizeModString(MOD_ID, GUI_WORK_ENERGY, "work energy line 1") {
                    +TextFormatting.GRAY
                    +energySystem.makeLightTextComponent(this@WorkEnergyIndicatorPiece.provider.workEnergyStored)
                    +energySystem.makeDarkTextComponent(this@WorkEnergyIndicatorPiece.provider.workEnergyCapacity)
                },
                localizeModString(MOD_ID, GUI_WORK_ENERGY, "work energy line 2") {
                    +TextFormatting.GRAY
                    +energySystem.makeLightTextComponent(this@WorkEnergyIndicatorPiece.provider.workEnergyTick)
                },
                localizeModString(MOD_ID, GUI_WORK_ENERGY, "work energy line 3") {
                    +TextFormatting.GRAY
                    +(this@WorkEnergyIndicatorPiece.provider.workEnergyCapacity / this@WorkEnergyIndicatorPiece.provider.workEnergyTick)
                        .makeTextComponent(TextFormatting.WHITE)
                }
            ) else mutableListOf(
                localizeModString(MOD_ID, GUI_WORK_ENERGY, "waiting for work") {
                    +TextFormatting.DARK_RED
                }
            )).map { it.formattedText }

            container.drawTooltip(lines, mouseX - guiX, mouseY - guiY)
        }
    }
}
